doc/source/guide/in_depth/functional_guide.rst

```diff
@@ -0,0 +1,257 @@
+
+Note that it is possible to use the class without fulfilling these assumtptions, however in this case some of the mathematical results might not hold.
+
+.. _convex conjugate: https://en.wikipedia.org/wiki/Convex_conjugate
+.. _Legendre transformation: https://en.wikipedia.org/wiki/Legendre_transformation
+
+.. _field: https://en.wikipedia.org/wiki/Field_(mathematics)
+.. _ordered field: https://en.wikipedia.org/wiki/Ordered_field
+.. _least-upper-bound property: https://en.wikipedia.org/wiki/Least-upper-bound_property
+
+.. _implementation:
+
+Implementation of functionals
+=============================
+
+To define your own functional, you start by writing::
+
+    class MyFunctional(odl.solvers.Functional):
+        ...
+
+Since `Functional` is a subclass of `Operator`, it has the *abstract method* ``domain`` which need to be explicitly overridden by any subclass.
+
+``domain``: `Set`
+    The domain of this functional, i.e., the set of elements to which this functional can be applied.
+
+Moreover, there are several optional parameters associated with a functional. These are
+
+``linear``: `bool`, optional
+    If `True`, the functional is considered as linear.
+    Default: ``False``
+``smooth``: `bool`, optional
+    If `True`, assume that the functional is continuously differentiable.
+    Default: ``False``
+``concave``: `bool`, optional
+    If `True`, assume that the functional is concave.
+    Default: ``False``
+``convex``: `bool`, optional
+    If `True`, assume that the functional is convex.
+    Default: ``False``
+``grad_lipschitz``: `float`, optional
+    The Lipschitz constant of the gradient.
+    Default: infinity
+
+A functional also has a number of optional methods and properties associated with it.
+The default value of these are all to raise a `NotImplemetedError`.
+The properties are:
+
+ * ``gradient``. This returns the gradient operator of the functional, i.e., the operator that corresponds to the mapping
+
+   .. math::
+
+      x \to \nabla f(x),
+
+   where :math:`\nabla f(x)` is the element used to evaluate the (directional) derivative in a direction :math:`d` by :math:`\langle \nabla f(x), d \rangle`.
+
+ * ``conjugate_functional``. This is the convex conjugate functional, also known as the Legendre transform or Fenchel conjugate.
+   It is defined as
+
+   .. math::
+
+      f^*(x^*) = \sup_{x \in X} \{ \langle x^*,x \rangle - f(x)  \},
+
+   where the element :math:`x^*` lives in the space :math:`X^*`, the (continuous/normed) dual space of :math:`X`, i.e., the space of all continuous linear functionals defined on :math:`X`.
+   Moreover, the notation :math:`\langle x^*,x \rangle` is the "dual pairing" between :math:`X` and :math:`X^*`, i.e., the evaluation of the linear functional :math:`x^*` in the point :math:`x`.
+   However, Hilbert spaces are self-dual, meaning :math:`X^* = X`, and :math:`\langle x^*,x \rangle` is then the inner product.
```

Comment:
Not sure this needs to be mentioned in this detail since we assume Hilbert spaces from the very beginning. Maybe just a short comment.